src/goto-instrument/reachability_slicer.cpp

```diff
@@ -54,11 +54,49 @@ void reachability_slicert::fixedpoint_to_assertions(
   const is_threadedt &is_threaded,
   slicing_criteriont &criterion)
 {
-  std::vector<cfgt::node_indext> src = get_sources(is_threaded, criterion);
+  // Stack entries are pairs of node indices and a boolean flag that indicates
+  // whether the function's callsite is known, in which case it has already
+  // been placed on the stack and function heads are a no-op.
 
-  std::vector<cfgt::node_indext> reachable = cfg.get_reachable(src, false);
-  for(const auto index : reachable)
-    cfg[index].reaches_assertion = true;
+  std::vector<std::pair<cfgt::node_indext, bool>> stack;
```

Comment:
Rather than having to introduce this comment I'd suggest you define a `struct` with members with meaningful names. That should make the code a lot easier to understand.